toolchain/check/context.cpp

```diff
@@ -216,6 +216,17 @@ auto Context::DiagnoseDuplicateName(SemIRLoc dup_def, SemIRLoc prev_def)
       .Emit();
 }
 
+auto Context::DiagnosePoisonedName(SemIRLoc loc) -> void {
+  // TODO: Improve the diagnostic to output the error where the name was
+  // poisoned and a note the name was later declared. See discussion in
+  // https://github.com/carbon-language/carbon-lang/pull/4654#discussion_r1876607172
+  CARBON_DIAGNOSTIC(
+      NameDeclUsedUnqualifiedBefore, Error,
+      "name previously used by unqualified name lookup and not found; "
+      "cannot later be declared");
+  emitter_->Build(loc, NameDeclUsedUnqualifiedBefore).Emit();
```

Comment:
```suggestion
  CARBON_DIAGNOSTIC(NameUseBeforeDecl, Error, "name used before it was declared");
  CARBON_DIAGNOSTIC(NameUseBeforeDeclNote, Note, "declared here");
  unit_info.emitter.Emit(Parse::NodeId::Invalid, NameUseBeforeDecl)
      .Note(loc, NameUseBeforeDeclNote)
      .Emit();
```

Reply:
Thanks! Done.
I was not aware that it's ok (for now) to put an invalid location, and that's why I thought we won't be able to error if we don't use the location we have.